source/material_model/rheology/compositional_viscosity_prefactors.cc

```diff
@@ -0,0 +1,170 @@
+#include <aspect/parameters.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    namespace Rheology
+    {
+      template <int dim>
+      CompositionalViscosityPrefactors<dim>::CompositionalViscosityPrefactors ()
+        = default;
+
+      template <int dim>
+      std::vector<double>
+      CompositionalViscosityPrefactors<dim>::compute_viscosities (const MaterialModel::MaterialModelInputs<dim> &in,
+                                                                  const double base_viscosity,
+                                                                  const unsigned int composition_index,
+                                                                  const unsigned int q) const
+      {
+        std::vector<double> factored_viscosities(number_of_prefactors, base_viscosity);
+        switch (viscosity_prefactor_scheme)
+          {
+            case none:
+            {
+              break;
+            }
+            case hk04_olivine_hydration:
+            {
+              // We calculate the atomic H/Si ppm (C_OH) at each point to compute the water fugacity of
+              // olivine assuming a composition of 90 wt% Forsterite and 10 wt% Fayalite from Hirth
+              // and Kohlstaedt 2004 10.1029/138GM06.
+              const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
+              const double weight_fraction_H2O = in.composition[q][bound_fluid_idx]; // mass fraction of bound water
+              const double weight_fraction_olivine = 1 - weight_fraction_H2O; // mass fraction of olivine
+              const double COH = (weight_fraction_H2O/molar_mass_H2O) / (weight_fraction_olivine/molar_mass_olivine) * 1e6; // COH in H / Si ppm
+              const double point_water_fugacity = COH / A_H2O * \
+                                                  std::exp((activation_energy_H2O + in.pressure[q]*activation_volume_H2O)/ \
+                                                           (constants::gas_constant * in.temperature[q]));
+
+              factored_viscosities[0] = base_viscosity*std::pow(point_water_fugacity, -diffusion_water_fugacity_exponents[composition_index]);
+              factored_viscosities[1] = base_viscosity*std::pow(point_water_fugacity, -dislocation_water_fugacity_exponents[composition_index]);
+              break;
+            }
+          }
+        return factored_viscosities;
+      }
+
+
+      template <int dim>
+      void
+      CompositionalViscosityPrefactors<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.declare_entry("Number of prefactors", "2",
+                          Patterns::List(Patterns::Integer(1)),
+                          "The number of flow laws to apply viscosity multiplicative prefactors. "
+                          "Default value is 2, for diffusion creep and dislocation creep. ");
+
+        prm.declare_entry ("Water fugacity exponents for diffusion creep", "0.0",
```

Comment:
For this and the next parameter, you should clarify that these parameters are only used when the prefactor scheme is HK04 olivine hydration.